cannon/mipsevm/multithreaded/mips.go

```diff
@@ -321,8 +321,23 @@ func (m *InstrumentedState) doMipsStep() error {
 	}
 	m.state.StepsSinceLastContextSwitch += 1
 
-	//instruction fetch
-	insn, opcode, fun := exec.GetInstructionDetails(m.state.GetPC(), m.state.Memory)
+	pc := m.state.GetPC()
+	cacheIdx := pc / 4
+
+	var insn, opcode, fun uint32
+	if int(cacheIdx) < len(m.cached_decode) {
+		decoded := m.cached_decode[cacheIdx]
+		if decoded.insn != 0 {
+			insn, opcode, fun = decoded.insn, decoded.opcode, decoded.fun
+		} else {
+			// Fallback decode for uninitialized slot
+			insn, opcode, fun = exec.GetInstructionDetails(pc, m.state.Memory)
+			m.cached_decode[cacheIdx] = InstructionDetails{insn, opcode, fun}
```

Comment:
When `decoded.insn == 0`, this may indeed be a valid instruction (for `sll`). And since the entire `m.cached_decode` buffer is allocated, it's not possible for there to be uninitialized values in there.
So falling back here is unnecessary. It's OK to use the zero valued instruction as-is.

Reply:
Okay, I understand, unfortunately this logic is what enables single step tests to work. Are you saying to remove it, and then the singlestep tests can fail within the ci tests

Reply:
Yes, I think it should be removed. I'll take a closer look at why the tests are failing without it. While it makes sense to have that fallback to accomodate tests, it's awkward to do so. There should be a better way to fix tests rather than have that contrivance in the VM.

Reply:
I think the best way to do this would be to move calls to `testutil.StoreInstruction` to the VMFactory. So that when the initial VM state is constructed, memory is also preset with the instruction used for the unit test. But I don't recommend doing this in this PR as it would be a big refactor.

For now, I think what you've suggested is OK. To reduce boilerplate, I suggest tweaking `testutil.StoreInstruction` so it updates the cache of the specific instruction being stored. This can be done as a single item update, rather than a cache refresh, so don't accidentally slowdown tests again.